File: china_finance/public/js/voucher_preparation.js
frappe.provide("china_finance.preparation");

(() => {
	const api = async (method, args) => (await frappe.call({method: `china_finance.services.voucher_preparation.${method}`, args, type: "POST", freeze: true})).message;
	const esc = value => frappe.utils.escape_html(String(value ?? ""));
	const money = (value, currency) => frappe.format(value, {fieldtype: "Currency", options: "currency"}, {only_value: true}, {currency});

	china_finance.preparation.edit = async (name, refresh) => {
		const result = await api("get_draft", {name});
		if (!result.can_edit) return frappe.msgprint(esc(result.reason));
		const doc = result.doc;
		const account_query = () => ({filters: {company: doc.company, is_group: 0, disabled: 0}});
		const fields = [
			{fieldname: "source_row", fieldtype: "Data", hidden: 1},
			{fieldname: "account", fieldtype: "Link", options: "Account", label: __("科目"), reqd: 1, in_list_view: 1, columns: 3, get_query: account_query},
			{fieldname: "user_remark", fieldtype: "Small Text", label: __("摘要"), in_list_view: 1, columns: 3},
			{fieldname: "debit_in_account_currency", fieldtype: "Float", precision: 2, label: __("借方（科目币种）"), in_list_view: 1, columns: 2},
			{fieldname: "credit_in_account_currency", fieldtype: "Float", precision: 2, label: __("贷方（科目币种）"), in_list_view: 1, columns: 2},
			{fieldname: "exchange_rate", fieldtype: "Float", label: __("汇率"), default: 1},
			{fieldname: "party_type", fieldtype: "Link", options: "DocType", label: __("往来类型"), get_query: () => ({filters: {name: ["in", ["Customer", "Supplier", "Employee", "Shareholder"]]}})},
			{fieldname: "party", fieldtype: "Dynamic Link", options: "party_type", label: __("往来单位")},
			{fieldname: "cost_center", fieldtype: "Link", options: "Cost Center", label: __("成本中心"), get_query: () => ({filters: {company: doc.company, is_group: 0}})},
			{fieldname: "project", fieldtype: "Link", options: "Project", label: __("项目")},
		];
		const dialog = new frappe.ui.Dialog({title: `${__("编辑未记账凭证")} · ${name}`, size: "extra-large", fields: [
			{fieldtype: "HTML", options: `<p>${esc(doc.company)} · ${esc(result.currency)} · ${__("保存后需要重新核对。银行收支必须与原回单一致。")}</p>`},
			{fieldname: "posting_date", fieldtype: "Date", label: __("凭证日期"), default: doc.posting_date, reqd: 1},
			{fieldname: "user_remark", fieldtype: "Small Text", label: __("凭证摘要"), default: doc.user_remark},
			{fieldname: "accounts", fieldtype: "Table", label: __("会计分录"), in_place_edit: true, fields,
				data: doc.accounts.map(row => ({...row, source_row: row.name}))},
			{fieldtype: "HTML", options: `${result.receipts.map(r => `<a href="${esc(r.source_file)}#page=${r.page_number}" target="_blank" rel="noopener">原回单：第 ${r.page_number} 页第 ${r.position} 张</a>`).join(" · ")}${result.cash.error ? `<p class="text-warning">${esc(result.cash.error)}</p>` : ""}`},
			{fieldname: "cash_flow_rows", fieldtype: "Table", label: __("现金流项目（按本位币核对）"), hidden: !result.cash.rows.length, cannot_add_rows: true, cannot_delete_rows: true, in_place_edit: true,
				data: result.cash.rows, fields: [
					{fieldname: "cash_account", fieldtype: "Data", label: __("现金或银行科目"), read_only: 1, in_list_view: 1, columns: 4},
					{fieldname: "direction", fieldtype: "Data", label: __("收付"), read_only: 1, in_list_view: 1, columns: 1},
					{fieldname: "amount", fieldtype: "Float", label: __("金额"), read_only: 1, in_list_view: 1, columns: 2},
					{fieldname: "row_code", fieldtype: "Select", label: __("现金流项目"), options: [{value: "", label: "请选择"}, ...result.cash.options], in_list_view: 1, columns: 3},
				]},
			{fieldname: "reason", fieldtype: "Small Text", label: __("修改说明（调整日期必填）")},
		], primary_action_label: __("保存草稿"), primary_action: async values => {
			dialog.get_primary_btn().prop("disabled", true);
			try {
				await api("save_draft", {name, modified: doc.modified, ...values});
				dialog.hide();
				frappe.show_alert({message: __("草稿已保存，请重新核对"), indicator: "green"});
				refresh?.();
			} finally { dialog.get_primary_btn().prop("disabled", false); }
		}});
		dialog.add_custom_action(__("原件与完整凭证"), () => frappe.set_route("Form", "Journal Entry", name));
		dialog.add_custom_action(__("预览打印（未记账）"), () => {
			const body = `<h3>未记账凭证 · ${esc(doc.posting_date)}</h3><p>${esc(doc.company)} · ${esc(result.currency)} · ${esc(name)}</p><table border="1" cellspacing="0" cellpadding="8"><tr><th>摘要</th><th>科目</th><th>借方</th><th>贷方</th></tr>${doc.accounts.map(r => `<tr><td>${esc(r.user_remark || doc.user_remark)}</td><td>${esc(r.account)}</td><td>${money(r.debit, result.currency)}</td><td>${money(r.credit, result.currency)}</td></tr>`).join("")}</table><p>未记账，仅供核对。此预览为打开时已保存的版本。</p>`;
			const win = window.open("", "_blank");
			if (win) { win.document.write(`<html><head><title>未记账凭证</title></head><body>${body}</body></html>`); win.document.close(); win.print(); }
		});
		dialog.show();
	};

	china_finance.preparation.trial = async (company, from_date, to_date) => {
		const result = await api("trial_balance", {company, from_date, to_date});
		const labels = {opening: "期初余额", posted_debit: "已记账借方", posted_credit: "已记账贷方", draft_debit: "草稿借方", draft_credit: "草稿贷方", expected_balance: "预计余额"};
		const dialog = new frappe.ui.Dialog({title: __("含未记账凭证的科目余额试算"), size: "extra-large", fields: [{fieldtype: "HTML", options: `
			<p>${esc(company)} · ${esc(from_date)} — ${esc(to_date)} · ${esc(result.currency)}</p><p>${esc(result.message)}</p>
			<p>本期草稿 ${result.draft_count} 张；更早期间草稿 ${result.earlier_drafts} 张。</p>
			${result.excluded.length ? `<div class="alert alert-warning">以下草稿未纳入，试算不完整：${result.excluded.map(r => `${esc(r.name)}：${esc(r.error)}`).join("<br>")}</div>` : ""}
			<div style="overflow:auto"><table class="table table-bordered"><thead><tr><th>科目</th>${Object.values(labels).map(label => `<th>${label}</th>`).join("")}</tr></thead><tbody>${result.rows.map(row => `<tr><td>${esc(row.account)}</td>${Object.keys(labels).map(key => `<td class="text-right">${money(row[key], result.currency)}</td>`).join("")}</tr>`).join("")}</tbody></table></div>`}]});
		dialog.show();
	};

	china_finance.preparation.review = async names => {
		const preview = await api("review_preview", {names});
		const options = row => Object.fromEntries(row.cash.options.map(o => [o.value, o.label]));
		const body = preview.map(row => `<details open><summary>${esc(row.name)} · ${esc(row.posting_date)}</summary><table class="table table-bordered"><tr><th>摘要</th><th>科目</th><th>借方</th><th>贷方</th></tr>${row.accounts.map(r => `<tr><td>${esc(r.summary)}</td><td>${esc(r.account)}</td><td>${esc(r.debit)}</td><td>${esc(r.credit)}</td></tr>`).join("")}</table><p>${row.cash.rows.map(r => `${esc(r.cash_account)} ${esc(r.direction)} ${esc(r.amount)} → ${esc(options(row)[r.row_code] || "待选择现金流项目")}`).join("<br>")}${esc(row.cash.error)}</p></details>`).join("");
		return new Promise(resolve => {
			const dialog = new frappe.ui.Dialog({title: __("核对凭证与现金流项目"), size: "extra-large", fields: [{fieldtype: "HTML", options: body}], primary_action_label: __("核对完成，待记账"), primary_action: async () => {
				dialog.get_primary_btn().prop("disabled", true);
				try {
					const rows = await api("review_drafts", {names, versions: Object.fromEntries(preview.map(r => [r.name, r.modified])), cash_plans: Object.fromEntries(preview.map(r => [r.name, r.cash.rows]))});
					frappe.msgprint(rows.map(r => `${esc(r.name)}：${r.success ? "核对完成，待记账" : esc(r.error)}`).join("<br>"));
					dialog.hide(); resolve(rows);
				} finally { dialog.get_primary_btn().prop("disabled", false); }
			}});
			dialog.onhide = () => resolve([]);
			dialog.show();
		});
	};

	china_finance.preparation.bind_report = report => {
		report.page.add_menu_item(__("凭证准备模式设置"), () => frappe.set_route("Form", "China Finance Settings", report.get_filter_value("company")));
		report.page.add_inner_button(__("核对所选草稿"), async () => {
			const indices = report.datatable?.rowmanager.getCheckedRows() || [];
			const names = [...new Set(indices.map(i => report.data[i]?.draft_name).filter(Boolean))];
			if (!names.length) return frappe.msgprint(__("请勾选未记账凭证的分录行；同张凭证只处理一次"));
			await china_finance.preparation.review(names);
			report.refresh();
		});
		report.page.add_inner_button(__("未记账试算"), () => china_finance.preparation.trial(report.get_filter_value("company"), report.get_filter_value("from_date"), report.get_filter_value("to_date")));
		report.page.add_inner_button(__("月末处理"), () => frappe.new_doc("China Closing Run", {company: report.get_filter_value("company"), from_date: report.get_filter_value("from_date"), to_date: report.get_filter_value("to_date")}));
		report.page.add_inner_button(__("新增草稿"), () => frappe.new_doc("Journal Entry", {company: report.get_filter_value("company")}));
	};
})();

File: china_finance/services/voucher_preparation.py
"""Editable voucher preparation; official ledgers are written only on submission."""

import hashlib
import json
from collections import defaultdict

import frappe
from frappe import _
from frappe.utils import cint, flt, getdate, now_datetime

READY_FIELDS = ("custom_china_ready_hash", "custom_china_ready_by", "custom_china_ready_on")
REVIEW_ROLES = ("System Manager", "Accounts Manager", "China Finance Manager", "China Voucher Reviewer")
HEAD_FIELDS = (
	"posting_date",
	"voucher_type",
	"company",
	"finance_book",
	"multi_currency",
	"is_opening",
	"user_remark",
	"cheque_no",
	"cheque_date",
	"custom_china_bank_transaction",
	"custom_china_cash_flow_plan",
)
LINE_FIELDS = (
	"account",
	"account_currency",
	"exchange_rate",
	"debit",
	"credit",
	"debit_in_account_currency",
	"credit_in_account_currency",
	"party_type",
	"party",
	"cost_center",
	"project",
	"reference_type",
	"reference_name",
	"reference_detail_no",
	"is_advance",
	"user_remark",
)
EDIT_FIELDS = (
	"account",
	"exchange_rate",
	"debit_in_account_currency",
	"credit_in_account_currency",
	"party_type",
	"party",
	"cost_center",
	"project",
	"user_remark",
)


def check_company(company):
	frappe.get_doc("Company", company).check_permission("read")


def mode_enabled(company, posting_date=None):
	from china_finance.services.voucher import get_company_settings

	settings = get_company_settings(company)
	return bool(
		settings
		and settings.get("enable_voucher_preparation")
		and settings.get("preparation_from_date")
		and (not posting_date or getdate(posting_date) >= getdate(settings.preparation_from_date))
	)


@frappe.whitelist()
def get_mode(company, posting_date=None):
	check_company(company)
	return {"enabled": mode_enabled(company, posting_date)}


def content_hash(doc):
	# Include configured accounting dimensions as well as the normal editable fields.
	from erpnext.accounts.doctype.accounting_dimension.accounting_dimension import get_accounting_dimensions

	dimensions = get_accounting_dimensions() or []

	def values(row, fields):
		result = {}
		for field in fields:
			value = row.get(field)
			df = row.meta.get_field(field)
			if df and df.fieldtype in ("Currency", "Float", "Percent", "Int", "Check"):
				value = flt(value)
			result[field] = value or None
		return result

	payload = {
		"header": values(doc, HEAD_FIELDS + tuple(dimensions)),
		"accounts": [values(row, LINE_FIELDS + tuple(dimensions)) for row in doc.accounts],
	}
	return hashlib.sha256(json.dumps(payload, sort_keys=True, default=str).encode()).hexdigest()


def is_ready(doc):
	return bool(doc.get("custom_china_ready_hash") and doc.custom_china_ready_hash == content_hash(doc))


def ensure_open(doc):
	from china_finance.services.source_voucher_edit import _get_closing_blockers

	blockers = _get_closing_blockers(doc)
	if blockers:
		frappe.throw("；".join(blockers))


def validate_preparation(doc, method=None):
	if doc.doctype != "Journal Entry" or not doc.meta.has_field(READY_FIELDS[0]):
		return
	old = doc.get_doc_before_save()
	# Read-only fields must not be forgeable through import, REST or a changed form payload.
	for field in READY_FIELDS:
		doc.set(field, old.get(field) if old else None)
	if doc.docstatus == 0:
		if mode_enabled(doc.company, doc.posting_date):
			ensure_open(doc)
		if old and old.get("custom_china_ready_hash") and not is_ready(doc):
			for field in READY_FIELDS:
				doc.set(field, None)


def guard_submission(doc, method=None):
	if doc.doctype == "Journal Entry" and mode_enabled(doc.company, doc.posting_date):
		if not is_ready(doc):
			frappe.throw(_("凭证尚未核对，或核对后内容已改变。请在查凭证中重新核对后记账。"))


def draft_documents(company, from_date, to_date):
	check_company(company)
	return [
		frappe.get_doc("Journal Entry", row.name)
		for row in frappe.get_list(
			"Journal Entry",
			filters={"company": company, "docstatus": 0, "posting_date": ["between", [from_date, to_date]]},
			fields=["name"],
			order_by="posting_date, creation, name",
			limit_page_length=0,
		)
	]


def validate_draft(doc):
	doc.check_permission("read")
	if doc.docstatus != 0:
		frappe.throw(_("{0} 已经记账或作废，请刷新").format(doc.name))
	ensure_open(doc)
	if not doc.accounts or abs(sum(flt(r.debit) - flt(r.credit) for r in doc.accounts)) > 0.005:
		frappe.throw(_("{0} 分录为空或借贷不平衡").format(doc.name))
	# Run native validation on an in-memory copy. Do not mutate the stored draft on review.
	doc.validate()
	from china_finance.services.bank_receipt_import import validate_linked_voucher

	validate_linked_voucher(doc)


@frappe.whitelist(methods=["POST"])
def review_drafts(names, versions=None, cash_plans=None):
	frappe.only_for(REVIEW_ROLES)
	names = frappe.parse_json(names) if isinstance(names, str) else names
	versions = frappe.parse_json(versions) if isinstance(versions, str) else versions
	cash_plans = frappe.parse_json(cash_plans) if isinstance(cash_plans, str) else cash_plans
	if not isinstance(versions, dict):
		frappe.throw(_("请刷新并核对凭证版本后再确认"))
	if not isinstance(names, list) or not names or len(names) > 300:
		frappe.throw(_("一次请选择 1 至 300 张草稿"))
	results = []
	for name in dict.fromkeys(names):
		point = "review_" + frappe.generate_hash(length=8)
		frappe.db.savepoint(point)
		try:
			doc = frappe.get_doc("Journal Entry", name)
			from china_finance.services.month_end import guard_period_write

			guard_period_write(doc)
			doc = frappe.get_doc("Journal Entry", name, for_update=True)
			doc.check_permission("write")
			if str(doc.modified) != versions.get(doc.name):
				frappe.throw(_("凭证已被修改，请刷新并重新核对"))
			if not mode_enabled(doc.company, doc.posting_date):
				frappe.throw(_("请先在中国财务设置中启用凭证准备模式并设置起始日期"))
			validate_draft(doc)
			cash = cash_plan(doc)
			if cash["error"]:
				frappe.throw(cash["error"])
			if cash["rows"]:
				if not isinstance(cash_plans, dict) or cash_plans.get(doc.name) != cash["rows"]:
					frappe.throw(_("现金流建议已变化，请刷新核对预览"))
				if any(not r["row_code"] for r in cash["rows"]):
					frappe.throw(_("请先在凭证编辑面板选择现金流项目"))
				doc.custom_china_cash_flow_plan = json.dumps(cash["rows"], sort_keys=True, ensure_ascii=False)
				doc.save(ignore_version=False)
			# Review an actual saved version, not any transient defaults set by validation.
			doc.reload()
			doc.db_set(
				{
					READY_FIELDS[0]: content_hash(doc),
					READY_FIELDS[1]: frappe.session.user,
					READY_FIELDS[2]: now_datetime(),
				}
			)
			doc.add_comment("Comment", _("已核对当前凭证内容，待记账；后续修改将要求重新核对。"))
			results.append({"name": name, "success": True})
		except Exception as exc:
			frappe.db.rollback(save_point=point)
			results.append({"name": name, "success": False, "error": str(exc)})
	return results


@frappe.whitelist()
def review_preview(names):
	names = frappe.parse_json(names) if isinstance(names, str) else names
	if not isinstance(names, list) or not 1 <= len(names) <= 300:
		frappe.throw(_("请选择 1 至 300 张凭证"))
	rows = []
	for name in dict.fromkeys(names):
		doc = frappe.get_doc("Journal Entry", name)
		doc.check_permission("read")
		rows.append(
			{
				"name": name,
				"modified": str(doc.modified),
				"cash": cash_plan(doc),
				"posting_date": doc.posting_date,
				"total_debit": doc.total_debit,
				"total_credit": doc.total_credit,
				"accounts": [
					{"account": r.account, "summary": r.user_remark, "debit": r.debit, "credit": r.credit}
					for r in doc.accounts
				],
			}
		)
	return rows


@frappe.whitelist()
def get_draft(name):
	doc = frappe.get_doc("Journal Entry", name)
	doc.check_permission("read")
	if doc.docstatus != 0:
		frappe.throw(_("该凭证已记账，请刷新后按受控更正流程处理"))
	from china_finance.services.source_voucher_edit import _build_status

	status = _build_status(doc)
	from china_finance.services.bank_receipt_import import get_voucher_receipts

	return {
		"doc": doc.as_dict(),
		"can_edit": status["can_edit"],
		"reason": status["reason"],
		"cash": cash_plan(doc),
		"receipts": get_voucher_receipts("Journal Entry", name),
		"currency": frappe.get_cached_value("Company", doc.company, "default_currency"),
	}


@frappe.whitelist(methods=["POST"])
def save_draft(
	name, modified, accounts, posting_date=None, user_remark=None, reason=None, cash_flow_rows=None
):
	doc = frappe.get_doc("Journal Entry", name)
	from china_finance.services.month_end import guard_period_write

	guard_period_write(doc)
	doc = frappe.get_doc("Journal Entry", name, for_update=True)
	doc.check_permission("write")
	if doc.docstatus != 0 or str(doc.modified) != str(modified):
		frappe.throw(_("凭证已经改变，请刷新后重新修改"), frappe.TimestampMismatchError)
	ensure_open(doc)
	accounts = frappe.parse_json(accounts) if isinstance(accounts, str) else accounts
	if not isinstance(accounts, list) or not 2 <= len(accounts) <= 300:
		frappe.throw(_("请填写 2 至 300 行有效分录"))
	old_rows = {r.name: r.as_dict() for r in doc.accounts}
	seen = set()
	doc.set("accounts", [])
	for row in accounts:
		key = row.get("source_row")
		if key and (key not in old_rows or key in seen):
			frappe.throw(_("分录来源已变化，请刷新凭证"))
		seen.add(key)
		data = old_rows.get(key, {}).copy()
		data.pop("idx", None)
		data.update({k: row.get(k) for k in EDIT_FIELDS if k in row})
		if data.get("account") != old_rows.get(key, {}).get("account"):
			data["account_currency"] = frappe.db.get_value("Account", data.get("account"), "account_currency")
		doc.append("accounts", data)
	if posting_date and getdate(posting_date) != getdate(doc.posting_date):
		if not (reason or "").strip():
			frappe.throw(_("调整凭证日期必须填写原因"))
		doc.posting_date = posting_date
	if user_remark is not None:
		doc.user_remark = user_remark
	if cash_flow_rows is not None:
		cash_flow_rows = (
			frappe.parse_json(cash_flow_rows) if isinstance(cash_flow_rows, str) else cash_flow_rows
		)
		doc.custom_china_cash_flow_plan = json.dumps(cash_flow_rows, sort_keys=True, ensure_ascii=False)
	doc.save(ignore_version=False)
	if reason:
		doc.add_comment("Comment", frappe.utils.escape_html(reason))
	return get_draft(name)


def cash_plan(doc):
	from china_finance.services.cash_equivalent_scope import get_cash_scope_accounts
	from china_finance.services.cash_flow_assignment import (
		INFLOW_ROWS,
		OUTFLOW_ROWS,
		_suggestion_by_account,
		_valid_cash_flow_rows,
		is_direct_assignment_required,
	)

	result = {"rows": [], "options": [], "error": ""}
	if not is_direct_assignment_required(doc.company, doc.posting_date):
		return result
	try:
		cash_accounts = set(get_cash_scope_accounts(doc.company, doc.posting_date))
		counterparts = defaultdict(float)
		cash = defaultdict(float)
		for r in doc.accounts:
			amount = flt(r.debit) - flt(r.credit)
			if r.account in cash_accounts and amount:
				cash[(r.account, "收款" if amount > 0 else "付款")] += abs(amount)
			else:
				counterparts[r.account] += amount
		counterparts = {k: v for k, v in counterparts.items() if abs(v) > 0.005}
		if not cash or not counterparts:
			return result
		codes, template = _valid_cash_flow_rows(doc.company, doc.posting_date)
		result["options"] = [
			{"value": r.row_code, "label": r.label} for r in template.rows if r.row_code in codes
		]
		suggestion = _suggestion_by_account(doc.company, doc.posting_date).get(
			max(counterparts, key=lambda k: abs(counterparts[k]))
		)
		saved = frappe.parse_json(doc.get("custom_china_cash_flow_plan") or "[]")
		if not isinstance(saved, list):
			frappe.throw(_("草稿现金流项目格式错误"))
		for (account, direction), amount in sorted(cash.items()):
			matching = [
				r
				for r in saved
				if r.get("cash_account") == account
				and r.get("direction") == direction
				and abs(flt(r.get("amount")) - amount) < 0.005
			]
			code = (
				matching[0].get("row_code")
				if len(matching) == 1
				else (
					suggestion.get("cash_inflow_row_code" if direction == "收款" else "cash_outflow_row_code")
					if suggestion
					else None
				)
			)
			valid = INFLOW_ROWS if direction == "收款" else OUTFLOW_ROWS
			if code not in codes or code not in valid | {"FX_EFFECT"}:
				code = None
			result["rows"].append(
				{"cash_account": account, "direction": direction, "amount": flt(amount, 2), "row_code": code}
			)
	except frappe.ValidationError as exc:
		result["error"] = str(exc)
	return result


def promote_cash_plan(doc, confirm=True):
	"""Apply the reviewed draft classification to actual GL legs and validate normally."""
	from china_finance.services.cash_flow_assignment import (
		confirm_cash_flow_assignment,
		create_assignment_if_required,
	)

	plan = frappe.parse_json(doc.get("custom_china_cash_flow_plan") or "[]")
	if not plan:
		return
	voucher = frappe.db.get_value(
		"China Accounting Voucher", {"source_key": "Posting|Journal Entry|" + doc.name}, "name"
	)
	name = create_assignment_if_required(voucher)
	if not name:
		return
	assignment = frappe.get_doc("China Cash Flow Assignment", name)
	if assignment.docstatus == 1:
		return
	assignment.check_permission("write")
	groups = defaultdict(float)
	for row in assignment.items:
		groups[(row.cash_account, row.cash_direction)] += flt(row.cash_amount)
	for row in assignment.items:
		matched = [
			r
			for r in plan
			if r["cash_account"] == row.cash_account
			and r["direction"] == row.cash_direction
			and abs(flt(r["amount"]) - groups[(row.cash_account, row.cash_direction)]) < 0.005
		]
		if len(matched) != 1:
			frappe.throw(_("实际现金分录与已核对草稿不同，请核对现金流项目"))
		row.cash_flow_row_code = matched[0]["row_code"]
		row.assigned_amount = row.cash_amount
	assignment.save()
	# Normal confirmation preserves role separation and amount/direction validation.
	if confirm:
		confirm_cash_flow_assignment(name)


def get_draft_ledger_rows(filters):
	if filters.get("voucher_status") in ("已记账", "已冲销", "Posted", "Reversed"):
		return []
	if filters.get("source_doctype") and filters.source_doctype != "Journal Entry":
		return []
	rows = []
	for doc in draft_documents(filters.company, filters.from_date, filters.to_date):
		if filters.get("source_name") and filters.source_name != doc.name:
			continue
		if filters.get("voucher_number") and filters.voucher_number != doc.name:
			continue
		if filters.get("accounting_period") and str(doc.posting_date)[:7] != filters.accounting_period:
			continue
		ready = is_ready(doc)
		if filters.get("voucher_status") == "待记账" and not ready:
			continue
		if filters.get("voucher_status") == "未记账" and ready:
			continue
		if filters.get("voucher_word"):
			continue  # Draft identifiers are never official voucher numbers.
		for row in doc.accounts:
			if any(filters.get(f) and filters[f] != row.get(f) for f in ("account", "party_type", "party")):
				continue
			remarks = row.user_remark or doc.user_remark or ""
			if (
				filters.get("search_text")
				and filters.search_text.lower() not in f"{doc.name} {remarks} {row.account}".lower()
			):
				continue
			rows.append(
				frappe._dict(
					voucher_snapshot="draft:" + doc.name,
					draft_name=doc.name,
					posting_date=doc.posting_date,
					accounting_period=str(doc.posting_date)[:7],
					statutory_number="草稿 · " + doc.name,
					source_doctype="Journal Entry",
					source_name=doc.name,
					voucher_status=3 if ready else 0,
					currency=frappe.get_cached_value("Company", doc.company, "default_currency"),
					entry_idx=row.idx,
					account=row.account,
					party_type=row.party_type,
					party=row.party,
					remarks=remarks,
					debit=row.debit,
					credit=row.credit,
					base_total_amount=flt(row.debit) + flt(row.credit),
				)
			)
	return rows


@frappe.whitelist()
def trial_balance(company, from_date, to_date):
	"""A consistent read of actual GL plus eligible drafts, in company currency."""
	check_company(company)
	frappe.has_permission("GL Entry", "read", throw=True)
	if not from_date or not to_date or getdate(from_date) > getdate(to_date):
		frappe.throw(_("请填写有效的试算期间"))
	# Acquire the same company lock used by all source writes; use fresh current reads.
	from china_finance.services.month_end import lock_company

	lock_company(company)
	docs = draft_documents(company, from_date, to_date)
	all_names = frappe.get_all(
		"Journal Entry",
		filters={"company": company, "docstatus": 0, "posting_date": ["between", [from_date, to_date]]},
		pluck="name",
	)
	if len(docs) != len(all_names):
		frappe.throw(_("没有读取本期全部凭证的权限，不能生成完整试算"), frappe.PermissionError)
	accounts = {
		r.name: r
		for r in frappe.get_list(
			"Account", filters={"company": company}, fields=["name", "account_number"], limit_page_length=0
		)
	}
	ledger = frappe.db.sql(
		"""SELECT account, posting_date, debit, credit FROM `tabGL Entry`
		WHERE company=%s AND posting_date<=%s AND is_cancelled=0 FOR UPDATE""",
		(company, to_date),
		as_dict=True,
	)
	visible_count = frappe.get_list(
		"GL Entry",
		filters={"company": company, "posting_date": ["<=", to_date], "is_cancelled": 0},
		fields=[{"COUNT": "name", "as": "total"}],
	)[0].total
	if visible_count != len(ledger):
		frappe.throw(_("没有读取完整总账范围的权限，不能生成完整试算"), frappe.PermissionError)
	if any(r.account not in accounts for r in ledger) or any(
		r.account not in accounts for d in docs for r in d.accounts
	):
		frappe.throw(_("没有读取全部会计科目的权限"), frappe.PermissionError)
	values = defaultdict(
		lambda: {"opening": 0, "posted_debit": 0, "posted_credit": 0, "draft_debit": 0, "draft_credit": 0}
	)
	for row in ledger:
		item = values[row.account]
		if getdate(row.posting_date) < getdate(from_date):
			item["opening"] += flt(row.debit) - flt(row.credit)
		else:
			item["posted_debit"] += flt(row.debit)
			item["posted_credit"] += flt(row.credit)
	excluded = []
	for doc in docs:
		try:
			validate_draft(doc)
		except Exception as exc:
			excluded.append({"name": doc.name, "error": str(exc)})
			continue
		for row in doc.accounts:
			values[row.account]["draft_debit"] += flt(row.debit)
			values[row.account]["draft_credit"] += flt(row.credit)
	rows = []
	for account, amounts in sorted(values.items()):
		rows.append(
			{
				"account": account,
				**{k: flt(v, 2) for k, v in amounts.items()},
				"expected_balance": flt(
					amounts["opening"]
					+ amounts["posted_debit"]
					- amounts["posted_credit"]
					+ amounts["draft_debit"]
					- amounts["draft_credit"],
					2,
				),
			}
		)
	earlier = frappe.db.count(
		"Journal Entry", {"company": company, "docstatus": 0, "posting_date": ["<", from_date]}
	)
	return {
		"rows": rows,
		"excluded": excluded,
		"draft_count": len(docs),
		"earlier_drafts": earlier,
		"currency": frappe.get_cached_value("Company", company, "default_currency"),
		"message": "试算包含本期有效草稿；期初仅含已记账数据。早期草稿须先处理，正式报表以实际记账数据为准。",
	}
